python/ray/tune/trial.py

```diff
@@ -283,6 +283,9 @@ def should_stop(self, result):
         if result.get(DONE):
             return True
 
+        if callable(self.stopping_criterion):
+            return self.stopping_criterion(self.trial_id, result)
```

Comment:
One last odd thing about this is that this trial object are checkpointed/pickled individually (for global checkpointing), and this causes the `stopping_criterion` to be checkpointed N times separately. 

A couple options:

1. remove self.stopping_criterion from `__getstate__`, and deal with this upon restore
2. move `should_stop` into `TrialRunner` and let the TrialRunner take care of the stopping criterion.

Reply:
Ah I see. This actually applies to a bunch of other attributes too, right? For example, 
`config`, `sync_to_driver_fn`,  `export_formats` etc... basically anything that's common to the entire `Experiment`.

In which case I'd go with the first option and remove all of these fields from `__getstate__` and separately persist `Experiment` prior to trial execution (just once). Then if each trial knows what experiment it belongs to it can restore that data from there. Does that make sense?

Reply:
Yeah that makes sense!